source/agora/common/crypto/Schnorr.d

```diff
@@ -0,0 +1,413 @@
+/*******************************************************************************
+
+    Schnorr signature implementation using Ed25519
+
+    TODO:
+      - Add extended description
+      - Compress signature
+      - Check side channels attack on equality comparison (?)
+      - Move `Scalar` and `Point` to libsodiumd (?)
+      - Expose the keypair in a "nicer" format
+      - Add test for more than 2 party
+      - Provide nicer APIs
+      - Work on the `R` interaction
+
+    Super TODO:
+      - Audit for more attacks
+      - Audit GDC-generated code (/LDC?)
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.crypto.Schnorr;
+
+import agora.common.Data;
+import agora.common.Hash;
+
+import geod24.bitblob;
+import libsodium;
+
+import std.algorithm;
+import std.range;
+
+
+/// Single signature example
+nothrow @nogc unittest
+{
+    Pair kp = Pair.random();
+    auto signature = sign(kp, "Hello world");
+    assert(verify(kp.V, signature, "Hello world"));
+}
+
+/// Multi-signature example
+nothrow @nogc unittest
+{
+    // Setup
+    static immutable string secret = "BOSAGORA for the win";
+    Pair kp1 = Pair.random();
+    Pair kp2 = Pair.random();
+    Pair R1 = Pair.random();
+    Pair R2 = Pair.random();
+    Point R = R1.V + R2.V;
+    Point X = kp1.V + kp2.V;
+
+    auto sig1 = sign(kp1.v, X, R, R1.v, secret);
+    auto sig2 = sign(kp2.v, X, R, R2.v, secret);
+    auto sig3 = Signature(R, sig1.s + sig2.s);
+
+    // No one can verify any of those individually
+    assert(!verify(kp1.V, sig1, secret));
+    assert(!verify(kp1.V, sig2, secret));
+    assert(!verify(kp2.V, sig2, secret));
+    assert(!verify(kp2.V, sig1, secret));
+    assert(!verify(kp1.V, sig3, secret));
+    assert(!verify(kp2.V, sig3, secret));
+
+    // But multisig works
+    assert(verify(X, sig3, secret));
+}
+
+/// Batch verification
+unittest
+{
+    enum BatchSize = 42;
+
+    Pair[BatchSize] keys = void;
+    for (size_t i = 0; i < keys.length; ++i)
+        keys[i] = Pair.random();
+
+    Hash[BatchSize] secrets = void;
+    Signature[BatchSize] sigs = void;
+    foreach (idx, ref key; keys)
+    {
+        secrets[idx] = hashFull(idx);
+        sigs[idx] = sign(key, secrets[idx]);
+    }
+
+    assert(verifyBatch(sigs[], keys[].map!(v => v.V), secrets[]));
+}
+
+/// TODO: Write in compressed form
+public struct Signature
+{
+    /// Commitment
+    public Point R;
+    /// Proof
+    public Scalar s;
+}
+
+/// Represent the message to hash (part of `c`)
+private struct Message (T)
+{
+    public Point X;
```

Comment:
Is there any way to document these? X, R.

Reply:
I see they are documented inside of `sign`. Well ok since this is a private struct I guess it's not necessary.